### PR TITLE
Machine Config Proposal

### PR DESCRIPTION
**Pull request by csrwng**
Instead of repeating the proposed types in a doc, this proposal includes code changes to represent the initial proposed type definitions.

The goal is to come to a basic consensus on an initial implementation and generate some discussion around the open issues. Like anything else, this is expected to evolve as we evolve the implementation.

### CODE REVIEW

File: docs/proposal/machine-config.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+## AWSMachineProviderConfig
+Defines the spec portion of a Machine specific to AWS. It is meant to be embedded in the
+Machine.Spec.ProviderConfig field of a cluster API Machine.
+
+Following are design decisions around this type:
+
+- No AWS SDK types are included. For the following reasons:
+  - Clients would also need to have the AWS SDK vendored.
+  - The type would be tied to a specific AWS client version.
```
Comment (mvladev) [inline docs/proposal/machine-config.md]: This 100% correct and valid, but we'll provide only subset of configurations, which you can do, when launching a machine.

For example in our project we decided to more or less copy the structs which were used in aws-client-go. https://github.com/gardener/machine-controller-manager/blob/5976512b001db0b029a8982dc93e1065f099e4a7/pkg/apis/machine/v1alpha1/types.go#L763-L885 . I'm not 100% sure that our approach is 100% valid as well.

File: docs/proposal/machine-config.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
+# AWS Provider Machine Configuration and Status
+
+## Overview
+This proposal aims to provide an initial API for AWS machine configuration and status for the AWS machine actuator.
+Cluster configuration for the AWS provider is out of scope for this proposal.
+
+## AWSMachineProviderConfig
+Defines the spec portion of a Machine specific to AWS. It is meant to be embedded in the
```
Comment (mvladev) [inline docs/proposal/machine-config.md]: Are we focusing this API only for on-demand AWS instances? If we want to have spot instances support then we might want to include additional fields.

File: cloud/aws/providerconfig/types.go [hunk 3]
```diff
@@ -17,19 +17,112 @@ import (
...
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags map[string]string
+
+	// IAMInstanceProfile is the name of the IAM role to assign to the new instance
+	IAMInstanceProfile string
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: a LocalObjectReference only contains a string (Name). But it is the more semantically correct type given that we are referencing another object in storage. See https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#references-to-related-objects
You can find a few examples in the core types like this one:
https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/types.go#L1256

File: cloud/aws/providerconfig/types.go [hunk 4]
```diff
@@ -17,19 +17,112 @@ import (
...
+	// SubnetID is the id of the subnet to use for this instance
+	SubnetID string
+
+	// AvailabilityZone is the availability zone where the instance should be
+	// placed. Exammple: us-east-1c
+	AvailabilityZone string
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it's created.
+	LoadBalancerIDs []string
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: Excellent question. I'm not sure that it's doable for machines that are part of the control plane, where we run into a chicken and egg problem trying to communicate with the control plane. However, I'm starting to think that maybe that's the responsibility of the cluster actuator and not the machine actuator. For everything else,  I think the LoadBalancer type of service should work. Wdyt?

File: docs/proposal/machine-config.md (hunk 2 above)
Comment (csrwng) [reply docs/proposal/machine-config.md]: I would say that the we could initially focus on on-demand (because it's simpler), but we could eventually evolve to support spot instances.

File: docs/proposal/machine-config.md (hunk 1 above)
Comment (csrwng) [reply docs/proposal/machine-config.md]: Can we adopt the approach of including as much of the API as we need for what we initially support and as the actuator evolves add more if we need to?

**Comment by csrwng** [comment]
Thank you @mvladev for the review and great suggestions!

**Comment by csrwng** [comment]
@dgoodwin @abutcher @joelddiaz @twiest

File: cloud/aws/providerconfig/types.go [hunk 5]
```diff
@@ -14,22 +14,151 @@
...
+	Value string
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSClusterProviderConfig struct {
 	metav1.TypeMeta `json:",inline"`
 }
 
+// AWSMachineProviderStatus is the type that will be embedded in a Machine.Status.ProviderStatus field.
+// It containsk AWS-specific status information.
```
Comment (dgoodwin) [inline cloud/aws/providerconfig/types.go]: -k.

File: cloud/aws/providerconfig/types.go [hunk 6]
```diff
@@ -14,22 +14,151 @@
...
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
+// for an AWS instance.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderConfig struct {
```
Comment (enxebre) [inline cloud/aws/providerconfig/types.go]: do we want to include an optional KeyName for ssh?

Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: definitely, thx!

Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: PR updated

File: cloud/aws/providerconfig/types.go [hunk 7]
```diff
@@ -14,22 +14,154 @@
...
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+
+	// Placement indicates where to create the instance in AWS
+	Placement Placement
```
Comment (ingvagabund) [inline cloud/aws/providerconfig/types.go]: I like this.

File: docs/proposal/machine-config.md (hunk 1 above)
Comment (frobware) [reply docs/proposal/machine-config.md]: I think this is really the best and most pragmatic choice.

File: cloud/aws/providerconfig/types.go [hunk 8]
```diff
@@ -14,22 +14,154 @@
...
+	PublicIP *bool
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+
+	// Placement indicates where to create the instance in AWS
```
Comment (frobware) [inline cloud/aws/providerconfig/types.go]: It's more than just "indication", no? How about `specifies where...`?

File: cloud/aws/providerconfig/types.go [hunk 9]
```diff
@@ -14,22 +14,154 @@
...
+	Placement Placement
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it's created.
```
Comment (frobware) [inline cloud/aws/providerconfig/types.go]: nit: "it is created"

File: cloud/aws/providerconfig/types.go [hunk 10]
```diff
@@ -14,22 +14,154 @@
...
+	LoadBalancerIDs []string
+}
+
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters
```
Comment (frobware) [inline cloud/aws/providerconfig/types.go]: How do you discriminate between the various choices? Which (of the optional) take precedence?

File: cloud/aws/providerconfig/types.go [hunk 11]
```diff
@@ -14,22 +14,154 @@
...
+	// ARN of resource
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
+}
+
+// Placement indicates where to create the instance in AWS
+type Placement struct {
+	// Region is the region to use to create the insntance
```
Comment (frobware) [inline cloud/aws/providerconfig/types.go]: typo: "create the instance".

File: cloud/aws/providerconfig/types.go [hunk 12]
```diff
@@ -14,22 +14,154 @@
...
+// AWSMachineProviderCondition is a condition in a AWSMachineProviderStatus
+type AWSMachineProviderCondition struct {
+	// Type is the type of the condition.
+	Type AWSMachineProviderConditionType
+	// Status is the status of the condition.
+	Status corev1.ConditionStatus
+	// LastProbeTime is the last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
```
Comment (frobware) [inline cloud/aws/providerconfig/types.go]: Is this "from one _state_ to another"?

File: cloud/aws/providerconfig/types.go [hunk 13]
```diff
@@ -14,22 +14,154 @@
...
+	// AvailabilityZone is the availability zone of the instance
+	AvailabilityZone string
+}
+
+// Filter is a filter used to identify an AWS resource
+type Filter struct {
+	// Name of the filter
+	Name string
+
+	// Values includes one or more filter values
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: It's coming from the EC2 AWS API which says the same thing. I don't believe that specifying a filter name without a value is a valid selector. So it should at least have one value.

File: cloud/aws/providerconfig/types.go (hunk 12 above)
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: It's from one status to the other ... referring to the Status field above... when it transitions from corev1.StatusTrue to corev1.StatusFalse or viceversa.

File: cloud/aws/providerconfig/types.go [hunk 14]
```diff
@@ -14,22 +14,156 @@
...
+	// CredentialsSecret is a reference to the secret with AWS credentials. Otherwise, defaults to permissions
+	// provided by attached IAM role where the actuator is running.
+	CredentialsSecret *corev1.LocalObjectReference
+
+	// KeyName is the name of the KeyPair to use for SSH
+	KeyName *string
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
```
Comment (mvladev) [inline cloud/aws/providerconfig/types.go]: this should be `*int` (optional)

File: cloud/aws/providerconfig/types.go [hunk 15]
```diff
@@ -14,22 +14,156 @@
...
+	// PublicIP specifies whether the instance should get a public IP. If not present,
+	// it should use the default of its subnet.
+	PublicIP *bool
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
```
Comment (errordeveloper) [inline cloud/aws/providerconfig/types.go]: For this to map to an ASG, then you would want a slice here.

File: cloud/aws/providerconfig/types.go [hunk 16]
```diff
@@ -14,22 +14,156 @@
...
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
+// for an AWS instance.
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSMachineProviderConfig struct {
```
Comment (errordeveloper) [inline cloud/aws/providerconfig/types.go]: Is this supposed to represent a single machine? What about ASGs?

Comment (mvladev) [reply cloud/aws/providerconfig/types.go]: @errordeveloper single machine - we won't use ASG

File: cloud/aws/providerconfig/types.go [hunk 17]
```diff
@@ -14,22 +14,156 @@
...
+// a validation error.
+type AWSResourceReference struct {
+	// ID of resource
+	ID *string
+
+	// ARN of resource
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
```
Comment (errordeveloper) [inline cloud/aws/providerconfig/types.go]: Is this mostly used for requests inputs?

Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: Yes

File: cloud/aws/providerconfig/types.go [hunk 18]
```diff
@@ -14,22 +14,156 @@
...
+	// Name of the tag
+	Name string
+
+	// Value of the tag
+	Value string
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 type AWSClusterProviderConfig struct {
 	metav1.TypeMeta `json:",inline"`
```
Comment (ingvagabund) [inline cloud/aws/providerconfig/types.go]: No need for the `json...` annotation part given all the data types are internal.

File: cloud/aws/providerconfig/types.go (hunk 15 above)
Comment (ingvagabund) [reply cloud/aws/providerconfig/types.go]: Can a cluster have more subnets? If so, do we want to allow it? The subnet(s) can be automatically detected based on cluster's vpc ID.

File: cloud/aws/providerconfig/types.go (hunk 14 above)
Comment (ingvagabund) [reply cloud/aws/providerconfig/types.go]: `ec2.InstanceNetworkInterfaceSpecification` wants `*int64`

**Comment by timothysc** [comment]
/assign @detiber 

Can you review please, and we can try to get @csrwng unblocked.

File: cloud/aws/providerconfig/types.go [hunk 19]
```diff
@@ -14,22 +14,156 @@
...
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
+
+	// IAMInstanceProfile is a reference to an IAM role to assign to the instance
+	IAMInstanceProfile *AWSResourceReference
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: I think this is outside the scope of the MVP

File: cloud/aws/providerconfig/types.go [hunk 20]
```diff
@@ -17,19 +17,112 @@ import (
...
+	// IAMInstanceProfile is the name of the IAM role to assign to the new instance
+	IAMInstanceProfile string
+
+	// UserDataSecret contains a local reference to a secret that contains the
+	// UserData to apply to the instance
+	UserDataSecret *corev1.LocalObjectReference
+
+	// CredentialsSecret is a reference to the secret with AWS credentials. Otherwise, defaults to permissions
+	// provided by attached IAM role where the actuator is running.
+	CredentialsSecret *corev1.LocalObjectReference
```
[the remark this answers is not in the document]
Comment (chuckha) [reply cloud/aws/providerconfig/types.go]: Each machine shouldn't need the aws credentials, just the actuator. I think we can remove this

File: cloud/aws/providerconfig/types.go [hunk 21]
```diff
@@ -14,22 +14,156 @@
...
+
+	// KeyName is the name of the KeyPair to use for SSH
+	KeyName *string
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: Is SSH required? I'd be in favor of nixing SSH until we *really* need it and I don't think we'll be needing it for the MVP

File: cloud/aws/providerconfig/types.go (hunk 14 above)
Comment (chuckha) [reply cloud/aws/providerconfig/types.go]: I think it would be cleaner to remove this for the MVP and add it back in later when we need it.

File: cloud/aws/providerconfig/types.go [hunk 22]
```diff
@@ -14,22 +14,156 @@
...
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
+
+	// PublicIP specifies whether the instance should get a public IP. If not present,
+	// it should use the default of its subnet.
+	PublicIP *bool
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: The MVP will be either entirely public or entirely private so each machine does not need this configuration

File: cloud/aws/providerconfig/types.go [hunk 23]
```diff
@@ -14,22 +14,156 @@
...
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance.
+	SecurityGroups []AWSResourceReference
+
+	// Subnet is a reference to the subnet to use for this instance
+	Subnet AWSResourceReference
+
+	// Placement specifies where to create the instance in AWS
+	Placement Placement
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: Placement is out of scope for MVP, this can go away for now

File: cloud/aws/providerconfig/types.go [hunk 24]
```diff
@@ -14,22 +14,156 @@
...
+
+	// LoadBalancerIDs is the IDs of the load balancers to which the new instance
+	// should be added once it is created.
+	LoadBalancerIDs []string
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: I think this will live on the cluster object which the machine actuator has access to

File: cloud/aws/providerconfig/types.go [hunk 25]
```diff
@@ -14,22 +14,156 @@
...
+}
+
+// AWSResourceReference is a reference to a specific AWS resource by ID, ARN, or filters.
+// Only one of ID, ARN or Filters may be specified. Specifying more than one will result in
+// a validation error.
+type AWSResourceReference struct {
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: This seems like it could get messy? How does this fit into the MVP?

File: cloud/aws/providerconfig/types.go [hunk 26]
```diff
@@ -14,22 +14,156 @@
...
+
+	// ARN of resource
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
+}
+
+// Placement indicates where to create the instance in AWS
+type Placement struct {
```
Comment (chuckha) [inline cloud/aws/providerconfig/types.go]: out of scope for MVP

**Comment by chuckha** [comment]
To give a little more context here, I'm more than happy to keep this PR as is if this work has resources, but right now it might be a little cumbersome for the agreed upon MVP. If you're happy with it and want to continue with the PR as is, we can definitely make it work.

**Comment by csrwng** [comment]
@chuckha Thank you for the review. I'm ok with paring it down if it'll reduce friction for an initial implementation and as we add features/function, add things back into it.

File: cloud/aws/providerconfig/types.go (hunk 25 above)
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: It allows you to specify references to things that might not exist yet when you create your machine/machineset config. For example, you may know what your security groups will be named, but you don't have id's for them yet. Your config can use a Filter for the name tag for each of your security groups.

**Comment by chuckha** [comment]
That sounds good to me -- we should track the stuff that we nix with a ticket. I expect the backlog to get properly populated with tasks for the MVP this week and I don't want us to forget to add things back

**Comment by csrwng** [comment]
@chuckha Addressed your comments. The Placement struct included the region for the machine. Are we assuming that in an initial implementation the actuator would have a default region that it uses?

**Comment by csrwng** [comment]
squashed and updated versioned types

File: cloud/aws/providerconfig/types.go [hunk 27]
```diff
@@ -14,25 +14,132 @@
...
+	AMI AWSResourceReference
+
+	// InstanceType is the type of instance to create. Example: m4.xlarge
+	InstanceType string
+
+	// Tags is the set of tags to add to apply to an instance, in addition to the ones
+	// added by default by the actuator. These tags are additive. The actuator will ensure
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
```
Comment (mvladev) [inline cloud/aws/providerconfig/types.go]: Isn't it better for this type to be `map[string]string`?
AWS client accepts `map[string]string` already and with this implementation, you have to loop over every single item to put them in a `map`

File: cloud/aws/providerconfig/types.go [hunk 28]
```diff
@@ -14,25 +14,132 @@
...
+	Status corev1.ConditionStatus
+	// LastProbeTime is the last time we probed the condition.
+	// +optional
+	LastProbeTime metav1.Time
+	// LastTransitionTime is the last time the condition transitioned from one status to another.
+	// +optional
+	LastTransitionTime metav1.Time
+	// Reason is a unique, one-word, CamelCase reason for the condition's last transition.
+	// +optional
+	Reason string
```
Comment (detiber) [inline cloud/aws/providerconfig/types.go]: Why wouldn't we bubble these up to the MachineStatus ErrorReason/ErrorMessage instead of having a separate Reason/Message here?

File: cloud/aws/providerconfig/types.go (hunk 27 above)
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: The only reason I did it this way is: https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#lists-of-named-subobjects-preferred-over-maps
But if there are no objections to reverting it back to a map, then I can do so.

Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: Actually now that I read the guidelines again... this is a pure map ... not a map of subobjects, so it doesn't make sense to make it an array.

File: cloud/aws/providerconfig/types.go [hunk 29]
```diff
@@ -14,25 +14,132 @@
...
+	// these tags are present, but will not remove any other tags that may exist on the
+	// instance.
+	Tags []TagSpecification
+
+	// IAMInstanceProfile is a reference to an IAM role to assign to the instance
+	IAMInstanceProfile *AWSResourceReference
+
+	// DeviceIndex is the index of the device on the instance for the network interface attachment.
+	// Defaults to 0.
+	DeviceIndex int
```
[the remark this answers is not in the document]
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: Sounds good, will remove

File: cloud/aws/providerconfig/types.go (hunk 28 above)
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: Conditions are meant to provide more detailed status, giving you time, condition, etc. and allowing the possibility of reporting multiple conditions at a time. If we want to provide a more concise status at the higher level, then I can add an ErrorReason and ErrorMessage there.

File: cloud/aws/providerconfig/types.go (hunk 27 above)
[the remark this answers is not in the document]
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: will do

File: cloud/aws/providerconfig/types.go (hunk 25 above)
Comment (csrwng) [reply cloud/aws/providerconfig/types.go]: So far all of the resources that we reference from the machine config have a corresponding Describe call that allows them to be looked up by the fields specified in the struct. I would say we could introduce separate types of references in the future if/when we need them. Is that ok?

Comment (detiber) [reply cloud/aws/providerconfig/types.go]: Sounds good, thanks for the clarification.

File: cloud/aws/providerconfig/types.go [hunk 30]
```diff
@@ -14,25 +14,132 @@
...
+
+	// ARN of resource
+	ARN *string
+
+	// Filters is a set of filters used to identify a resource
+	Filters []Filter
+}
+
+// Filter is a filter used to identify an AWS resource
+type Filter struct {
```
[the remark this answers is not in the document]
Comment (detiber) [reply cloud/aws/providerconfig/types.go]: Makes sense, can we add a link to that to the comments to make it more clear for new users?

File: cloud/aws/providerconfig/types.go (hunk 28 above)
Comment (detiber) [reply cloud/aws/providerconfig/types.go]: There is already an ErrorReason/ErrorMessage on the MachineStatus object: https://github.com/kubernetes-sigs/cluster-api/blob/051f338bdacb76117d73a86258bc58c946add7b5/pkg/apis/cluster/v1alpha1/machine_types.go#L112-L133

It sounds like there could be a use case for duplicating the Reason/Message here as well for more granularity, though.

File: cloud/aws/providerconfig/types.go (hunk 16 above)
Comment (errordeveloper) [reply cloud/aws/providerconfig/types.go]: I though this subproject wasn't going to be so prescriptive, and it would be up to any given flavour to make a decision.

File: cloud/aws/providerconfig/types.go [hunk 31]
```diff
@@ -14,25 +14,123 @@
 package providerconfig
 
 import (
+	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+// AWSMachineProviderConfig is the type that will be embedded in a Machine.Spec.ProviderConfig field
```
Comment (errordeveloper) [inline cloud/aws/providerconfig/types.go]: This doesn't actually say anything about what this struct is supposed to represent. Is it a single running instance, or template of an instance or what? The comment must clarify that, at the very least. Instead it talks about relationship between structs, which is something one can figure by reading the code anyway.

File: cloud/aws/providerconfig/types.go (hunk 15 above)
Comment (errordeveloper) [reply cloud/aws/providerconfig/types.go]: I think there confusion around what this struct represents, as I pointed out above, hence the meaning (and type) of `Subnet` field is unclear.

File: cloud/aws/providerconfig/types.go (hunk 17 above)
Comment (errordeveloper) [reply cloud/aws/providerconfig/types.go]: Perhaps, it'd make more sense to use a separate struct for inputs and outputs. That's what AWS SDK does and it's more convenient to not create a additional model that one has to thing about.

File: cloud/aws/providerconfig/types.go (hunk 30 above)
[the remark this answers is not in the document]
Comment (errordeveloper) [reply cloud/aws/providerconfig/types.go]: Why do we redefine the type here? Why not alias https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#Filter?

File: cloud/aws/providerconfig/types.go (hunk 31 above)
Comment (errordeveloper) [reply cloud/aws/providerconfig/types.go]: Also, I think it's very reasonable to ask where this should be just an alias to [`ec2.Instance`](https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#Instance)?

**Comment by csrwng** [comment]
@detiber @errordeveloper thank you for your comments. I've now updated the PR.

File: cloud/aws/providerconfig/types.go (hunk 16 above)
Comment (detiber) [reply cloud/aws/providerconfig/types.go]: @errordeveloper That is a goal for the overall project, yes. That said, I think there is an inherent impedance mismatch between a Machine and an ASG-managed instance that we probably need to solve more generically in the great cluster-api.

To me it feels like the MachineSet controller would be a better manager for ASG-backed instances rather than the Machine controller. I am interested on hearing more of your thoughts on this topic, though.

**Comment by csrwng** [comment]
@errordeveloper I mentioned the reasons for not using the EC2 types directly in the proposal doc. Having separate types also gives us some freedom to choose the EC2 client we use. For example, at some point we may want to switch to the [V2 client](https://aws.amazon.com/blogs/developer/aws-sdk-for-go-2-0-developer-preview/).

File: cloud/aws/providerconfig/types.go [hunk 32]
```diff
@@ -14,25 +14,140 @@
...
+	// 2. Cluster/flavor setting
+	// 3. Subnet default
+	// +optional
+	PublicIP *bool
+
+	// SecurityGroups is an array of references to security groups that should be applied to the
+	// instance. These security groups would be set in addition to any security groups defined
+	// at the cluster level or in the actuator.
+	// +optional
+	SecurityGroups []AWSResourceReference
```
Comment (detiber) [inline cloud/aws/providerconfig/types.go]: Can this be renamed to AdditionalSecurityGroups to align with the updated comments?

**Comment by csrwng** [comment]
done

**Comment by detiber** [comment]
/lgtm

@errordeveloper are you able to attend the implementation meeting (https://groups.google.com/d/msg/kubernetes-sig-cluster-lifecycle/__XIKigkxkA/j1BHBdznDwAJ https://docs.google.com/document/d/10dq54Fd-xa6P5Iy3p46VY1YTFqugGMd1PygDIpuRw6c/edit#) to discuss your questions/concerns further?

**Comment by errordeveloper** [comment]
I would like to attend regularly, but it clashes with general implementer's
office hours, which I am currently chairing. Let's discuss how meetings
align/misalign in the WG call today.

**Comment by errordeveloper** [comment]
As far as I am aware, the V2 client is not very different with respect to
these struct are implemented.
